### ISSUE TITLE
Version 4.x: unable to hash build context ... could not evaluate symlink

### ISSUE THREAD

**Issue by dominikzogg** [pulumi/pulumi-docker#541]
### What happened?

```
Diagnostics:
  docker:index:Image (other-project):
    error: unable to hash build context: error while hashing "project/.eslintrc.json": could not evaluate symlink atproject/.eslintrc.json: lstat project: no such file or directory
```

### Expected Behavior

Docker image could be built as it was built in 3.x

### Steps to reproduce

Add a symlink within the code of context

### Output of `pulumi about`

```
CLI          
Version      3.57.1
Go Version   go1.20.1
Go Compiler  gc

Plugins
NAME          VERSION
digitalocean  4.18.0
docker        4.0.0
kubernetes    3.24.1
nodejs        unknown
postgresql    3.6.0

Host     
OS       fedora
Version  37
Arch     x86_64

This project is written in nodejs: executable='/run/user/1000/fnm_multishells/6117_1678693753068/bin/node' version='v18.12.1'

.........

Dependencies:
NAME                  VERSION
@pulumi/kubernetes    3.24.1
@pulumi/postgresql    3.6.0
@pulumi/pulumi        3.57.1
@types/node           18.15.1
prettier              2.8.4
typescript            4.9.5
@pulumi/digitalocean  4.18.0
@pulumi/docker        4.0.0

Pulumi locates its logs in /tmp by default
```

### Additional context

This symlink references to another symlink (i need to have it like this)

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by AaronFriel** [comment]
Hey @dominikzogg, thanks for reporting. Would it be possible for you to create a minimal repro in a Git repository for us to evaluate?

I'm curious as well, if you CD into the directory that contains `Pulumi.yaml` and attempt to `cat project/.eslintrc.json`, does that succeed? Looking at the code we use, it recursively walks symlinks to determine the true path, and the error indicates that one of the symlinks may point at a file that doesn't exist.

**Comment by dominikzogg** [comment]
@AaronFriel I tried to strip it down as possbile. https://github.com/dominikzogg/demo . I practices there are many "project" directories. Thats the reason why its structured this way

**Comment by dominikzogg** [comment]
I haven't checked the output good enough. Its complaining in other-project for project within the repository.

```
Diagnostics:
  docker:index:Image (other-project):
    error: unable to hash build context: error while hashing "project/.eslintrc.json": could not evaluate symlink at project/.eslintrc.json: lstat project: no such file or directory
```

**Comment by praneetloke** [comment]
I've just hit this issue as well with v4. It seems that having the context be the root of the repo causes some sort of a problem with symlinks in the `node_modules` folder for me. However, it doesn't fail when I narrow-down the `context` to the folder where the `Dockerfile` exists, even with a `node_modules` folder in it. I'll try to get a repro later.

**Comment by AaronFriel** [comment]
@dominikzogg Thanks for the repro, that made debugging this a snap. CC @praneetloke on that.

This should be fixed in the next release, pending merge of #544.

**Comment by dominikzogg** [comment]
@AaronFriel thanks for fixing, works fine
